bycycle/features/shape.py

```diff
@@ -0,0 +1,213 @@
+        - ``sample_last_trough`` : sample of the last trough
+        - ``sample_next_trough`` : sample of the next trough
+
+    Notes
+    -----
+    Peak vs trough centering
+
+        - By default, the first extrema analyzed will be a peak, and the final one a trough.
+        - In order to switch the preference, the signal is simply inverted and columns are renamed.
+        - Columns are slightly different depending on if ``center_extrema`` is set to 'peak' or
+          'trough'.
+
+    """
+
+    # Set defaults if user input is None
+    if find_extrema_kwargs is None:
+        find_extrema_kwargs = {'filter_kwargs': {'n_cycles': 3}}
+
+    elif 'first_extrema' in find_extrema_kwargs.keys():
+        raise ValueError('''This function has been designed to assume that the first extrema
+            identified will be a peak. This cannot be overwritten at this time.''')
+
+    # Negate signal if set to analyze trough-centered cycles
+    if center_extrema == 'peak':
+        pass
+    elif center_extrema == 'trough':
+        sig = -sig
+    else:
+        raise ValueError('Parameter "center_extrema" must be either "peak" or "trough"')
+
+    # Find extrema and zero-crossings locations in the signal
+    peaks, troughs = find_extrema(sig, fs, f_range, **find_extrema_kwargs)
+    rises, decays = find_zerox(sig, peaks, troughs)
+
+    # For each cycle, identify the sample of each extrema and zero-crossing
+    df_samples = compute_samples(peaks, troughs, decays, rises)
+
+    # Initialize dictionary for shape features
+    shape_features = {}
+
+    # Compute durations of period, peaks, and troughs
+    shape_features['period'] = df_samples['sample_next_trough'] - df_samples['sample_last_trough']
+    shape_features['time_peak'] = df_samples['sample_zerox_decay'] - df_samples['sample_zerox_rise']
+    shape_features['time_trough'] = rises - decays[:-1]
+
+    # Determine extrema voltage
+    shape_features['volt_peak'] = sig[peaks[1:]]
+    shape_features['volt_trough'] = sig[troughs[:-1]]
+
+    # Determine rise and decay characteristics
+    shape_features['time_decay'] = (troughs[1:] - peaks[1:])
+    shape_features['time_rise'] = (peaks[1:] - troughs[:-1])
+
+    shape_features['volt_decay'] = sig[peaks[1:]] - sig[troughs[1:]]
+    shape_features['volt_rise'] = sig[peaks[1:]] - sig[troughs[:-1]]
+    shape_features['volt_amp'] = (shape_features['volt_decay'] + shape_features['volt_rise']) / 2
+
+    # Compute rise-decay symmetry features
+    shape_features['time_rdsym'] = shape_features['time_rise'] / shape_features['period']
+
+    # Compute peak-trough symmetry features
+    shape_features['time_ptsym'] = shape_features['time_peak'] / \
+            (shape_features['time_peak'] + shape_features['time_trough'])
+
+    # Compute average oscillatory amplitude estimate during cycle
+    amp = amp_by_time(sig, fs, f_range, hilbert_increase_n=hilbert_increase_n, n_cycles=3)
```

Comment:
Here, n_cycles should use a controllable input, I think? Maybe from `find_extrema_kwargs`?

Reply:
Fixed this (a953843).